plugin/code_actions.py

```diff
@@ -63,52 +84,215 @@ def get_location_key(self, view: sublime.View, point: int) -> str:
 actions_manager = CodeActionsManager()
 
 
-def request_code_actions(view: sublime.View, point: int,
-                         actions_handler: Callable[[CodeActionsByConfigName], None]) -> CodeActionsAtLocation:
-    diagnostics_by_config = filter_by_point(view_diagnostics(view), Point(*view.rowcol(point)))
-    return request_code_actions_with_diagnostics(view, diagnostics_by_config, actions_handler)
+def request_code_actions(
+    view: sublime.View,
+    actions_handler: Callable[[CodeActionsByConfigName], None],
+    point: int,
+) -> CodeActionsCollector:
+    actions_collector = CodeActionsCollector(actions_handler)
+    with actions_collector:
+        file_name = view.file_name()
+        if file_name:
+            diagnostics_by_config = filter_by_point(view_diagnostics(view), Point(*view.rowcol(point)))
+            sessions = sessions_for_view(view, 'codeActionProvider')
+            for session in sessions:
+                config_name = session.config.name
+                if config_name in diagnostics_by_config:
+                    diagnostics = diagnostics_by_config[config_name]
+                    request = Request.codeAction(_create_code_action_request_params(view, file_name, diagnostics))
+                    session.send_request(request, actions_collector.create_collector(config_name))
+    return actions_collector
+
+
+def request_code_actions_on_save(
+    view: sublime.View,
+    actions_handler: Callable[[CodeActionsByConfigName], None],
+    on_save_actions: Dict[str, bool]
+) -> CodeActionsCollector:
+    def filtering_collector(
+        config_name: str,
+        kinds: List[str],
+        actions_collector: CodeActionsCollector
+    ) -> Tuple[Callable[[CodeActionsResponse], None], Callable[[Any], None]]:
+        """
+        Filters actions returned from the session so that only matching kinds are collected.
+
+        Since older servers don't support the "context.only" property, these will return all
+        actions that need to be filtered.
+        """
+        def actions_filter(actions: CodeActionsResponse) -> List[CodeActionOrCommand]:
+            return [a for a in (actions or []) if a.get('kind') in kinds]
+        collector = actions_collector.create_collector(config_name)
+        return (
+            lambda actions: collector(actions_filter(actions)),
+            lambda error: collector([])
+        )
+
+    actions_collector = CodeActionsCollector(actions_handler)
+    with actions_collector:
```

Comment:
It took me a while to understand why we need to have this `with` statement and why we need the `__enter__` and `__exit__` magic methods.

Reply:
I could add some comments in the implementation of `__enter__` or the collector class itself.

The previous implementation relied solely on synchronous execution to know how many responses are expected. With `with` it's now explicit that when the context is exited, we have triggered all requests by then.

Also, for the same reason, previous implementation did not work at all for sync implementation of code actions.

Reply:
Added explanation in the documentation block.